test/cpp_api_parity/module_impl_check.py

```diff
@@ -0,0 +1,289 @@
+# The purpose of this test is to check that we have implementation parity between
+# a Python `torch.nn` module and its corresponding C++ `torch::nn` module. Concretely,
+# this test does the following:
+#
+# 1. Get a test params dict from common_nn.py, run forward and backward on the
+# Python module created using the test params.
+#
+# 2. Serialize the Python module's parameters / buffers and its forward input
+# arguments, deserialize them in C++ and load them into the C++ module.
+#
+# 3. Run the same forward and backward passes on the C++ module, and serialize
+# the C++ module's forward output and backward gradients.
+#
+# 4. Compare Python/C++ module's forward output and backward gradients. If they
+# are the same, then we have implementation parity between Python/C++ module.
+
+import tempfile
+import shutil
+from string import Template
+import types
+
+import torch
+from cpp_api_parity.utils import TorchNNModuleTestParams, TORCH_NN_COMMON_TEST_HARNESS, \
+    compile_cpp_code_inline, set_python_tensors_requires_grad, move_python_tensors_to_device, \
+    add_test, compute_cpp_args_construction_stmts_and_forward_arg_symbols, serialize_arg_dict_as_script_module, \
+    compute_arg_dict, decorate_test_fn, compute_temp_file_path, generate_error_msg
+from cpp_api_parity import torch_nn_modules
+
+# Expected substitutions:
+#
+# ${module_variant_name}
+# ${module_qualified_name}
+# ${cpp_tmp_folder}
+# ${cpp_args_construction_stmts}
+# ${cpp_constructor_args}
+# ${device}
+# ${cpp_forward_args_symbols}
+TORCH_NN_MODULE_TEST_FORWARD_BACKWARD = Template("""
+void ${module_variant_name}_test_forward_backward(
+    const std::string& arg_dict_file_path,
+    const std::string& module_file_path,
+    const std::string& forward_output_file_path,
+    const std::string& backward_grad_dict_file_path) {
+  pybind11::gil_scoped_release no_gil;
+
+  // Declare arguments
+  auto arg_dict = load_dict_from_file(arg_dict_file_path);
+  ${cpp_args_construction_stmts};
+
+  // Construct module and load params/buffers from Python module
+  ${module_qualified_name} module${cpp_constructor_args};
+  module->to(std::string("${device}"));
+  torch::load(module, module_file_path);
+
+  // Some modules (such as `RReLU`) create random tensors in their forward pass.
+  // To make sure the random tensors created are the same in Python/C++, we need
+  // to set the RNG seed manually.
+  torch::manual_seed(0);
+
+  // Forward pass
+  auto cpp_output = module(${cpp_forward_args_symbols});
+
+  // Save the output into a file to be compared in Python later
+  write_ivalue_to_file(torch::IValue(cpp_output), forward_output_file_path);
+
+  // Backward pass
+  cpp_output.sum().backward();
+
+  // Put all gradients into a c10::Dict, save it into a file to be compared in Python later
+  c10::Dict<std::string, torch::Tensor> grad_dict;
+  for (const auto& param : module->named_parameters()) {
+    torch::Tensor grad = param.value().grad();
+    if (grad.is_sparse()) {
```

Comment:
This doesn't really help us check parity between if the python implementation is giving sparse gradients and if the c++ implementation is giving sparse gradients right?

We don't have a lot of modules that do this (is embedding the only one?) so this isn't blocking but it would be good to note somewhere

Reply:
Thanks for the catch! Fixed it by serializing `.indices()` and `.values()` separately for sparse tensors, so that we are able to check parity between if the python implementation is giving sparse gradients and if the c++ implementation is giving sparse gradients

Reply:
nice! 😅